### DISCUSSION TITLE
Postgres/MySQL in production

### DISCUSSION THREAD

**Discussion by plainlystated** [BretFisher/ama#148] [Q&A] [answered]
What's the current thinking on running postgres/mysql in production, with respects to things like:
- ways to avoid data loss/corruption?
- any concerns about contention with other roles/containers on the same node?
- any specific config suggestions for databases running in containers?

**Comment by BretFisher** [comment, marked as answer]
See my [docker production talk](https://dockercon.docker.com/watch/WdAeLaLuSCNQwEp61YVXUt) which talks about testing your performance and how many containers in a host (which are treated just like regular programs in most performance traits) is different than one app per VM, and many VM's on a host.

In Swarm, you can reserve and limit resources for services, so I recommend you do that for databases, which are designed to use all the memory you give them. If you ran them on a VM (container or not) with other apps, they will usually starve the other apps of memory unless you put limits in place. If you're using Swarm, you usually don't want things running on your managers, unless it's just a hobby project or test/ci system. It's best for ease of management and general performance to keep the managers isolated so a busy website or db doesn't cause Swarm leader election or raft db replication to fail.

You want your database containers to store their db/logs/index files in volumes, which the official images for psql/mysql do by default. Those volumes are just files on the host under `/var/lib/docker/volumes` so you could have another container running backup jobs against the servers themselves (dbdump) and copying them off the server.

In general, if you're using a major cloud vendor, and they have a DB service, use it. They are better at running DB's then 99% of us, and you'll get a lot more for your money then trying to run a real prod db solution with HA, monitoring, logging, backups, replication, etc.

Containers and container orchestration don't solve your DB HA issues really. If you need real uptime then you need a DB cluster/replica that you do in the DB's preferred way. Containers don't change that.

**Comment by plainlystated** [comment]
Great, thanks for thanks for the tips, especially the one about limiting resources. I'll give the video a watch and let you know if I have further questions :)
